funsor/adjoint.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import defaultdict
+import torch
+
+import funsor.ops as ops
+from funsor.interpreter import interpretation, reinterpret
+from funsor.ops import AssociativeOp
+from funsor.registry import KeyedRegistry
+from funsor.terms import Binary, Funsor, Number, Reduce, Variable, eager
+from funsor.torch import Tensor
+
+
+class AdjointTape(object):
+
+    def __init__(self):
+        self.tape = []
+
+    def __call__(self, cls, *args):
+        result = eager(cls, *args)
```

Comment:
Changing this from `reflect` to `eager` seems to have solved the issues I was seeing before.